### PR TITLE
feat(react-native): upgrade react-native to 0.68.1

### PR DESCRIPTION
**Pull request by xiongemi**
update react native to 0.68.1

## Current Behavior

- add projectRoot option for nx metro config, the projectRoot should be the app project folder
- also add the watchFolders option for nx metro config, in case developers want to watch a lib folder that is not under libs/
- known issue, e2e detox test with android does not work with RN 0.68.1, open github issue on detox: https://github.com/wix/Detox/issues/3338

---
to upgrade to react-native 68, develoers need to manually pull down the ios/android native code. it is not backwards compatible. however, to upgrade, easiest way to do:
### ios
- change the ios/Mobile/AppDelegate.m under app, line:
```
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"apps/studio-ghibli-search-engine-mobile/src/main" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

```
change it to relative path like src/main from apps/studio-ghibli-search-engine-mobile/src/main and remove the `fallbackResource:nil`.
The line would become:
```
return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"src/main.tsx"];
```

### android
if you got an jdk error, you might need to upgrade your jdk https://www.oracle.com/java/technologies/downloads/
```
ERROR: requires JDK11 or higher.
Incompatible major version detected: '8'
```
also, this one is not backwards compatible, the easies way is just create a new React Native app and copy the android folder over

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/react-native/src/generators/application/files/app/metro.config.js.template [hunk 1]
```diff
@@ -28,8 +28,12 @@ module.exports = (async () => {
       // Change this to true to see debugging info.
       // Useful if you have issues resolving modules
       debug: false,
-      // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx'
+      // all the file extensions used for imports other than 'ts', 'tsx', 'js', 'jsx', 'json'
       extensions: [],
+      // the project root to start the metro server
+      projectRoot: __dirname,
+      // Specify any additional (to projectRoot) watch folders, this is used to know which files to watch
```
Comment (jaysoo) [inline packages/react-native/src/generators/application/files/app/metro.config.js.template:35]: Wording could be better, maybe this?

```
// Specify folders to watch, in addition to Nx defaults (workspace libraries and node_modules)
```

File: packages/react-native/plugins/metro-resolver.ts [hunk 2]
```diff
@@ -136,17 +134,14 @@ function tsconfigPathsResolver(
...
 }
 
 let matcher: MatchPath;
 let absoluteBaseUrl: string;
 let paths: Record<string, string[]>;
 
-function getMatcher() {
-  const DEBUG = process.env.NX_REACT_NATIVE_DEBUG === 'true';
-
+function getMatcher(DEBUG: boolean) {
```
Comment (jaysoo) [inline packages/react-native/plugins/metro-resolver.ts:144]: Nitpick: Arguments and variables should be mixed case (`debug`).

File: packages/react-native/plugins/with-nx-metro.ts [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
+  watchFolders = watchFolders.concat([
+    join(workspaceRoot, 'node_modules'),
+    join(workspaceRoot, workspaceLayout().libsDir),
+    join(workspaceRoot, '.storybook'),
+  ]);
+  if (opts.watchFolders?.length) {
+    watchFolders = watchFolders.concat(opts.watchFolders);
+  }
+
+  watchFolders = watchFolders.filter((folder) => existsSync(folder));
```
Comment (jaysoo) [inline packages/react-native/plugins/with-nx-metro.ts:37]: Is there an issue if we don't filters? Like if the folder doesn't exist, but is added afterwards?

Probably just an edge case that requires restart?

Comment (xiongemi) [reply packages/react-native/plugins/with-nx-metro.ts:37]: yes. without filters, i got an error like `Error: ENOENT: no such file or directory, lstat '/Users/emilyxiong/Code/tmp/nx-react-native-68-1/.storybook'` if i create an app without storybook
it requires a restart when adding folders

File: packages/react-native/plugins/metro-resolver.ts (hunk 2 above)
Comment (xiongemi) [reply packages/react-native/plugins/metro-resolver.ts:144]: done

File: packages/react-native/plugins/metro-resolver.ts [hunk 4]
```diff
@@ -110,10 +108,10 @@ function tsconfigPathsResolver(
   extensions: string[],
   realModuleName: string,
   moduleName: string,
-  platform: string
+  platform: string,
+  DEBUG: boolean
```
[the remark this answers is not in the document]
Comment (xiongemi) [reply packages/react-native/plugins/metro-resolver.ts]: changed

File: packages/react-native/src/generators/application/files/app/metro.config.js.template (hunk 1 above)
Comment (xiongemi) [reply packages/react-native/src/generators/application/files/app/metro.config.js.template]: changed

**Pull request merged**